src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                return s_defaultAssemblies.Value;
+            }
         }
 
         private void WriteTypes(Assembly assembly)
         {
             WriteObject(assembly.GetTypes(), true);
         }
 
-        private void CompileSourceToAssembly(string source)
+        #endregion LoadAssembly
+
+        #region SourceCodeProcessing
+
+        private OutputKind OutputAssemblyTypeToOutputKind(OutputAssemblyType outputType)
+        {
+            switch (outputType)
+            {
+                case OutputAssemblyType.Library:
+                    return OutputKind.DynamicallyLinkedLibrary;
+                case OutputAssemblyType.ConsoleApplication:
+                    return OutputKind.ConsoleApplication;
+                case OutputAssemblyType.WindowsApplication:
+                    return OutputKind.WindowsApplication;
+                default:
+                    throw new ArgumentOutOfRangeException("outputType");
+            }
+        }
+
+        private CommandLineArguments DefaultArgsParse(IEnumerable<string> args)
         {
-            CSharpParseOptions parseOptions;
-            if (Language == Language.CSharp)
+            string sdkDirectory = s_defaultSdkDirectory;
+            string baseDirectory = this.SessionState.Path.CurrentLocation.Path;
+            string additionalReferenceDirectories = null;
+
+            switch (Language)
+            {
+                case Language.CSharp:
+                    return CSharpCommandLineParser.Default.Parse(args, baseDirectory, sdkDirectory, additionalReferenceDirectories);
+                case Language.VisualBasic:
+                    return VisualBasicCommandLineParser.Default.Parse(args, baseDirectory, sdkDirectory, additionalReferenceDirectories);
+                default:
+                    Diagnostics.Assert(false, "ParseText: Unsupported language family.");
+                    break;
+            }
+
+            return null;
+        }
+
+        private SyntaxTree ParseText(SourceText sourceText, ParseOptions parseOptions, string path = "")
+        {
+            switch (Language)
+            {
+                case Language.CSharp:
+                    return CSharpSyntaxTree.ParseText(sourceText, (CSharpParseOptions) parseOptions, path);
+
+                case Language.VisualBasic:
+                    return VisualBasicSyntaxTree.ParseText(sourceText, (VisualBasicParseOptions) parseOptions, path);
+
+                default:
+                    Diagnostics.Assert(false, "ParseText: Unsupported language family.");
+                    break;
+            }
+
+            return null;
+        }
+
+        private CompilationOptions GetDefaultCompilationOptions()
+        {
+            switch (Language)
+            {
+                case Language.CSharp:
+                    var csharpCompilationOptions = new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));
+
+                    return csharpCompilationOptions;
```

Comment:
Then, why not just `return new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));`?

Reply:
Fixed.